Bug Report: [JDK21] MultiRelease JAR in EAR (with Lucene)/FISH-7959

### Brief Summary

Hello,

We are checking our application (Jakarata EE10 with payara-embedded-all 6.2023.9) with the JDK21. All seems to run smoothly except for [Lucene](https://lucene.apache.org/), which does not find a class during its lookup:

```
Caused by: java.lang.ClassNotFoundException: org.apache.lucene.store.MemorySegmentIndexInputProvider
        at com.sun.enterprise.loader.ASURLClassLoader.findClassData(ASURLClassLoader.java:807)
        at com.sun.enterprise.loader.ASURLClassLoader.findClass(ASURLClassLoader.java:689)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:593)
        at com.sun.enterprise.loader.CurrentBeforeParentClassLoader.loadClass(CurrentBeforeParentClassLoader.java:83)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:534)
        at java.base/java.lang.Class.forName(Class.java:513)
        at java.base/java.lang.invoke.MethodHandles$Lookup.findClass(MethodHandles.java:2869)
        at org.apache.lucene.store.MMapDirectory.lookupProvider(MMapDirectory.java:422)
```

All the libraries related to Lucene are included in our EAR.

This works fine with the JDK17 because the *lucene-core* dependency was not a multi-release JAR before JDK19. However, since then there are some classes in `META-INF/versions` and the lookup in Payara fails.

```
META-INF/services/
META-INF/services/org.apache.lucene.analysis.TokenizerFactory
META-INF/services/org.apache.lucene.codecs.Codec
META-INF/services/org.apache.lucene.codecs.DocValuesFormat
META-INF/services/org.apache.lucene.codecs.KnnVectorsFormat
META-INF/services/org.apache.lucene.codecs.PostingsFormat
META-INF/services/org.apache.lucene.index.SortFieldProvider
META-INF/versions/
META-INF/versions/19/
META-INF/versions/19/org/
META-INF/versions/19/org/apache/
META-INF/versions/19/org/apache/lucene/
META-INF/versions/19/org/apache/lucene/store/
META-INF/versions/19/org/apache/lucene/store/MemorySegmentIndexInput$MultiSegmentImpl.class
META-INF/versions/19/org/apache/lucene/store/MemorySegmentIndexInput$SingleSegmentImpl.class
META-INF/versions/19/org/apache/lucene/store/MemorySegmentIndexInput.class
META-INF/versions/19/org/apache/lucene/store/MemorySegmentIndexInputProvider.class
META-INF/versions/20/
META-INF/versions/20/org/
META-INF/versions/20/org/apache/
META-INF/versions/20/org/apache/lucene/
META-INF/versions/20/org/apache/lucene/internal/
META-INF/versions/20/org/apache/lucene/internal/vectorization/
META-INF/versions/20/org/apache/lucene/internal/vectorization/PanamaVectorUtilSupport.class
META-INF/versions/20/org/apache/lucene/internal/vectorization/PanamaVectorizationProvider.class
META-INF/versions/20/org/apache/lucene/store/
META-INF/versions/20/org/apache/lucene/store/MemorySegmentIndexInput$MultiSegmentImpl.class
META-INF/versions/20/org/apache/lucene/store/MemorySegmentIndexInput$SingleSegmentImpl.class
META-INF/versions/20/org/apache/lucene/store/MemorySegmentIndexInput.class
META-INF/versions/20/org/apache/lucene/store/MemorySegmentIndexInputProvider.class
META-INF/versions/21/
META-INF/versions/21/org/
META-INF/versions/21/org/apache/
META-INF/versions/21/org/apache/lucene/
META-INF/versions/21/org/apache/lucene/store/
META-INF/versions/21/org/apache/lucene/store/MemorySegmentIndexInput$MultiSegmentImpl.class
META-INF/versions/21/org/apache/lucene/store/MemorySegmentIndexInput$SingleSegmentImpl.class
META-INF/versions/21/org/apache/lucene/store/MemorySegmentIndexInput.class
META-INF/versions/21/org/apache/lucene/store/MemorySegmentIndexInputProvider.class
```


### Expected Outcome

The lookup of the classes inside Lucene should work fine.

### Current Outcome

It crashes with a `ClassNotFoundException`.

### Reproducer

You will find a reproducer [here](https://github.com/ctabin/gf-test-lucene).

### Operating System

Linux Debian

### JDK Version

OpenJDK 21

### Payara Distribution

Payara Embedded All

--- a/nucleus/common/common-util/src/main/java/com/sun/enterprise/loader/ASURLClassLoader.java
+++ b/nucleus/common/common-util/src/main/java/com/sun/enterprise/loader/ASURLClassLoader.java
@@ -37,7 +37,7 @@
  * only if the new code is made subject to such option by the copyright
  * holder.
  */
- // Portions Copyright [2016-2020] [Payara Foundation and/or its affiliates]
+ // Portions Copyright [2016-2023] [Payara Foundation and/or its affiliates]
 
 package com.sun.enterprise.loader;
 
@@ -614,23 +614,13 @@ public class ASURLClassLoader extends CurrentBeforeParentClassLoader
             try {
                 if (res.isJar) { // It is a jarfile..
                     JarFile zip = res.zip;
-                    JarEntry entry = zip.getJarEntry(entryName);
+                    JarFile jar = new JarFile(res.file, false, JarFile.OPEN_READ,
+                            Runtime.Version.parse(System.getProperty("java.version")));
+                    JarEntry entry = jar.getJarEntry(entryName);
                     if (entry != null) {
-                        InputStream classStream = zip.getInputStream(entry);
-                        byte[] classData = getClassData(classStream);
+                        byte[] classData = getClassData(zip.getInputStream(entry));
                         res.setProtectionDomain(ASURLClassLoader.this, entry.getCertificates());
                         return classData;
-                    }
-                    
-                    if (zip.isMultiRelease()) {
-                        String javaVersion = System.getProperty("java.version");
-                        JarEntry multiVersionEntry = zip.getJarEntry("META-INF/versions/"+javaVersion+"/"+entryName);
-                        if (multiVersionEntry != null) {
-                            InputStream classStream = zip.getInputStream(multiVersionEntry);
-                            byte[] classData = getClassData(classStream);
-                            res.setProtectionDomain(ASURLClassLoader.this, multiVersionEntry.getCertificates());
-                            return classData;
-                        }
                     }
                 } else { // Its a directory....
                     String entryPath = entryName.replace('/', File.separatorChar);

--- a/nucleus/common/common-util/src/main/java/com/sun/enterprise/loader/ASURLClassLoader.java
+++ b/nucleus/common/common-util/src/main/java/com/sun/enterprise/loader/ASURLClassLoader.java
@@ -621,13 +621,37 @@ public class ASURLClassLoader extends CurrentBeforeParentClassLoader
                         res.setProtectionDomain(ASURLClassLoader.this, entry.getCertificates());
                         return classData;
                     }
+                    
+                    if (zip.isMultiRelease()) {
+                        String javaVersion = System.getProperty("java.version");
+                        JarEntry multiVersionEntry = zip.getJarEntry("META-INF/versions/"+javaVersion+"/"+entryName);
+                        if (multiVersionEntry != null) {
+                            InputStream classStream = zip.getInputStream(multiVersionEntry);
+                            byte[] classData = getClassData(classStream);
+                            res.setProtectionDomain(ASURLClassLoader.this, multiVersionEntry.getCertificates());
+                            return classData;
+                        }
+                    }
                 } else { // Its a directory....
-                    File classFile = new File (res.file, entryName.replace('/', File.separatorChar));
+                    String entryPath = entryName.replace('/', File.separatorChar);
+                    File classFile = new File (res.file, entryPath);
                     if (classFile.exists()) {
                         try (InputStream classStream = new FileInputStream(classFile)) {
                             byte[] classData = getClassData(classStream);
                             res.setProtectionDomain(ASURLClassLoader.this, null);
                             return classData;
+                        }
+                    }
+                    
+                    File multiVersionDir = new File(res.file, "META-INF/versions/");
+                    if (multiVersionDir.exists()) {
+                        File multiVersionClassFile = new File(multiVersionDir, entryPath);
+                        if (multiVersionClassFile.exists()) {
+                            try (InputStream classStream = new FileInputStream(multiVersionClassFile)) {
+                                byte[] classData = getClassData(classStream);
+                                res.setProtectionDomain(ASURLClassLoader.this, null);
+                                return classData;
+                            }
                         }
                     }
                 }

